src/components/tooltip/index.tsx

```diff
@@ -0,0 +1,118 @@
+import React, { FC, ReactNode, useState, useRef } from "react"
+
+import PopUp from "./popUp"
+
+export enum TooltipPosition {
+  top = "top",
+  topLeft = "topLeft",
+  topRight = "topRight",
+  bottom = "bottom",
+  bottomLeft = "bottomLeft",
+  bottomRight = "bottomRight",
+  left = "left",
+  right = "right",
+}
+
+const spacingPosition = {
+  top: "top",
+  topLeft: "top",
+  topRight: "top",
+  bottom: "bottom",
+  bottomLeft: "bottom",
+  bottomRight: "bottom",
+  left: "left",
+  right: "right",
+}
+
+const oppositePosition = {
+  top: TooltipPosition.bottom,
+  topLeft: TooltipPosition.bottomLeft,
+  topRight: TooltipPosition.bottomRight,
+  bottom: TooltipPosition.top,
+  bottomLeft: TooltipPosition.topLeft,
+  bottomRight: TooltipPosition.topRight,
+  left: TooltipPosition.right,
+  right: TooltipPosition.left,
+}
+
+const getSpaceAroundRect = ({ top, bottom, left, right }) => ({
+  top,
+  left,
+  bottom: window.innerHeight - bottom,
+  right: window.innerWidth - right,
+})
+
+export const calculateTooltipPosition = (
+  container,
+  originalPosition,
+  switchThreshold
+) => {
+  if (!switchThreshold && !container) {
+    return originalPosition
+  }
+
+  const space = getSpaceAroundRect(container.getBoundingClientRect())[
+    spacingPosition[originalPosition]
+  ]
+
+  return space < switchThreshold
+    ? oppositePosition[originalPosition]
+    : originalPosition
+}
+
+interface Props {
+  /**
+   * Content of the tooltip
+   */
+  renderContent: () => ReactNode
+  /**
+   * Position of the tooltip
+   */
+  position: TooltipPosition
+  /**
+   * Threshold of the position switch
+   * If the space around tooltip container (on the side where tooltip is to be displayed)
+   * is smaller than then threshold the tooltip will be moved
+   */
+  positionSwitchThreshold?: number
+}
+
+const Tooltip: FC<Props> = ({
+  children,
+  renderContent,
```

Comment:
What do we gain from a render props function in this case when we don't pass any functions into it like we could straight up replace it even without a custom hook and it would do the same?
As from [an article](https://kentcdodds.com/blog/when-to-not-use-render-props) regarding render props you sent me a while ago
> UPDATE: If you can use react@16.8.0 then the answer is to rarely use render props and almost always use a custom hook. HOOKS ARE ALMOST ALWAYS THE SUPERIOR PATTERN. Read about when you might still use render props

Reply:
I want to pass a tooltip content and not render it. And since I don't want to have a very awkward markup (like the library you linked) I went for render prop. That's one of the cases that can't be covered by a hook.